src/mavsdk/plugins/telemetry/telemetry_impl.cpp

```diff
@@ -1176,6 +1181,42 @@ void TelemetryImpl::process_battery_status(const mavlink_message_t& message)
     }
 }
 
+void TelemetryImpl::process_battery_status_v2(const mavlink_message_t& message)
+{
+    mavlink_battery_status_v2_t bat_status;
+    mavlink_msg_battery_status_v2_decode(&message, &bat_status);
+
+    _has_bat_status = true;
+
+    Telemetry::Battery new_battery;
+    new_battery.id = bat_status.id;
+    new_battery.voltage_v = bat_status.voltage; 
+    //Is it correct to set this to NaN for UINT32_MAX? HOw do you specify = NaN?
+
+    // FIXME: it is strange calling it percent when the range goes from 0 to 1.
+    // Can we fix this? What about if value not provided?
+    new_battery.remaining_percent = bat_status.percent_remaining; // to test * 1e-2f;
+
+    //bat_status also has:
+    //uint32_t current; /*< [mA] Battery current (through all cells/loads). UINT32_MAX: field not provided.*/
+    // uint32_t current_consumed; /*< [mAh] Consumed charge (since vehicle powered on). UINT32_MAX: field not provided. Note: For power modules the expectation is that batteries are fully charged before turning on the vehicle.*/
+    // uint32_t current_remaining; /*< [mAh] Remaining charge (until empty). UINT32_MAX: field not provided. Note: Power monitors should not set this value.*/
+    // uint32_t status_flags; /*<  Fault, health, and readiness status indications.*/
+    // int16_t temperature; /*< [cdegC] Temperature of the whole battery pack (not internal electronics). INT16_MAX field not provided.*/
+
+
+    set_battery(new_battery);
```

Comment:
Small concern here: is it the drone's responsibility to not send "conflicting" battery messages (I think there are 3 ways: battery_v1, battery_v2 (this one) and something going through the heartbeat (or a similar message)?). Or should MAVSDK somehow prioritize ("if I receive battery_v1, ignore the one in the heartbeat, if I receive battery_v2, ignore the heartbeat and battery_v1")?

Reply:
It isn't really a concern.

You should be able to map either or the two battery status messages to the same values. If a battery or the flight stack is sending out different values for the same battery in the same field purpose that would have to be a bug. There are some modes/faults (that we are not exposing yet) that might appear in one message and not the other, but not "clashes".

The SYS_STATUS (not heartbeat) is not indexed and that can be a problem. In a single-battery system all the messages should have battery values for the same name fields. But in a multi battery system the value of SYS_STATUS is "undefined". For MAVSDK we should never use that message for battery information. I haven't been able to deprecate it but we don't need it.

IMO we should take either v1 or v2 and use both if both come in. Why? Well, as I said there should be no clash. Further, I think most systems will only be outputting both messages for a small time, if at all. So there isn't much point putting in code to just use one or the other.

Reply:
Ok :+1:, thanks! :blush: